service/storage_proxy.cc

```diff
@@ -159,6 +165,85 @@ class storage_proxy::remote {
     netw::connection_drop_slot_t _connection_dropped;
     netw::connection_drop_registration_t _condrop_registration;
 
+    class request_info {
+    public:
+        virtual future<schema_ptr> get_schema() const = 0;
+        virtual dht::token get_token(const schema& schema) const = 0;
+    };
+
+    struct read_request_info: public request_info {
+        schema_ptr _schema;
+        const dht::partition_range& _range;
+    public:
+        read_request_info(schema_ptr schema, const dht::partition_range& range)
+            : _schema(schema)
+            , _range(range)
+        {
+        }
+        virtual future<schema_ptr> get_schema() const override {
+            return make_ready_future<schema_ptr>(_schema);
+        }
+        virtual dht::token get_token(const schema&) const override {
+            return end_token(_range);
+        }
+    };
+
+    class write_request_info: public request_info {
+        remote& _remote;
+        table_schema_version _schema_version;
+        netw::msg_addr _msg_from;
+        partition_key_view _partition_key;
+    public:
+        write_request_info(remote& remote, table_schema_version schema_version, netw::msg_addr msg_from,
+            partition_key_view _partition_key)
+            : _remote(remote)
+            , _schema_version(schema_version)
+            , _msg_from(msg_from)
+            , _partition_key(std::move(_partition_key))
+        {
+        }
+        virtual future<schema_ptr> get_schema() const override {
+            return _remote.get_schema_for_write(_schema_version, _msg_from);
+        }
+        virtual dht::token get_token(const schema& schema) const override {
+            return schema.get_partitioner().get_token(schema, _partition_key);
+        }
+    };
+
+    future<std::optional<replica::stale_topology_exception>> try_apply_fence(fencing_token fence,
+        lowres_clock::time_point timeout,
+        const request_info& request_info)
+    {
+        while (true) {
+            const auto current_version = _sp._shared_token_metadata.get()->get_topology_version();
+            if (fence.topology_version == current_version) {
+                co_return std::nullopt;
+            }
+            if (fence.topology_version < current_version) {
+                auto schema = co_await request_info.get_schema();
+                const auto& ks_name = schema->ks_name();
+                const auto& ks = _sp._db.local().find_keyspace(ks_name);
+                auto erm = ks.get_effective_replication_map();
+                const auto token = request_info.get_token(*schema);
```

Comment:
Not a big deal but can't part of those calculation be done only once and not on each iteration?

Reply:
But these calculations are done only once, we will either return null or throw if this `if` branch was taken.

Reply:
Right.